### DISCUSSION TITLE
pytorch classification | train_step and test_step error

### DISCUSSION THREAD

**Discussion by kunalsonalkar** [mrdbourke/pytorch-deep-learning#610] [Q&A]
```
def train_step(model: torch.nn.Module,
               data_loader: torch.utils.data.DataLoader,
               loss_fn: torch.nn.Module,
               optimizer: torch.optim.Optimizer,
               accuracy_fn):
    train_loss, train_acc = 0, 0
    #model.to(device)
    for batch, (X, y) in enumerate(data_loader):
        # Send data to GPU
        #X, y = X.to(device), y.to(device)

        # 1. Forward pass
        y_pred = model(X)

        # 2. Calculate loss
        loss = loss_fn(y_pred, y)
        train_loss += loss
        train_acc += accuracy_fn(y_true=y,
                                 y_pred=y_pred.argmax(dim=1)) # Go from logits -> pred labels

        # 3. Optimizer zero grad
        optimizer.zero_grad()

        # 4. Loss backward
        loss.backward()

        # 5. Optimizer step
        optimizer.step()

    # Calculate loss and accuracy per epoch and print out what's happening
    train_loss /= len(data_loader)
    train_acc /= len(data_loader)
    print(f"Train loss: {train_loss:.5f} | Train accuracy: {train_acc:.2f}%")

def test_step(data_loader: torch.utils.data.DataLoader,
              model: torch.nn.Module,
              loss_fn: torch.nn.Module,
              accuracy_fn):
    test_loss, test_acc = 0, 0
    #model.to(device)
    model.eval() # put model in eval mode
    # Turn on inference context manager
    with torch.inference_mode(): 
        for X, y in data_loader:
            # Send data to GPU
            #X, y = X.to(device), y.to(device)
            
            # 1. Forward pass
            test_pred = model(X)
            
            # 2. Calculate loss and accuracy
            test_loss += loss_fn(test_pred, y)
            test_acc += accuracy_fn(y_true=y,
                y_pred=test_pred.argmax(dim=1) # Go from logits -> pred labels
            )
        
        # Adjust metrics and print out
        test_loss /= len(data_loader)
        test_acc /= len(data_loader)
        print(f"Test loss: {test_loss:.5f} | Test accuracy: {test_acc:.2f}%\n")
```
I am trying to functionize my training and testing steps. They work well separately but when I am putting them into a function, the accuracy is going significantly down. Can someone please take a look at this? 
This was covered in 03. PyTorch Computer Vision. 

Thank you.

**Comment by shisirkha** [comment]
I think you are using the Same X and y variables for this instead of renaming the other to other names like X_test,y_test 

this might be overlapping 

Here the fixed code:
```
#Import tqdm for progress bar
from tqdm.auto import tqdm
def train_step(model: nn.Module,
              data_loader: torch.utils.data.DataLoader,
              loss_fn: torch.nn.Module,
              optimizer: torch.optim.Optimizer,
              accuracy_fn,
              device: torch.device = device):
  #Performs a training model trying to learn on data loader
  train_loss,train_acc = 0,0
  ###TRAIN TIME
  model = model.to(device)
  model.train()
  # Add a loop to loop through the training batches
  for batch,(X,y) in enumerate(data_loader):
    X,y = X.to(device),y.to(device)
    y_pred = model(X)
    loss = loss_fn(y_pred,y)
    train_loss += loss
    train_acc += accuracy_fn(y,y_pred.argmax(dim=1))
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

  #Print out what's happening
  #Device total train loss by length of train dataloader
  train_loss /= len(data_loader)
  train_acc /= len(data_loader)
  print(f"Train acc: {train_acc} | train_loss: {train_loss}")

def test_step(model: nn.Module,
              test_dataloader:  torch.utils.data.DataLoader,
              loss_fn: torch.nn.Module,
              accuracy_fn,
              device: torch.device = device):
  test_loss, test_acc = 0,0
  model = model.to(device)
  model.eval()
  with torch.inference_mode():
    for X_test, y_test in test_dataloader:
      X_test, y_test = X_test.to(device), y_test.to(device)
      test_pred = model(X_test)
      test_loss += loss_fn(test_pred,y_test)
      test_acc += accuracy_fn(y_test,test_pred.argmax(dim=1))
    #Calculate the test loss average per batch
    test_loss /= len(test_dataloader)
    #Calculate average accuracy per batch
    test_acc/=len(test_dataloader)
  #Print out what happening
  print(f"\ntest loss: {test_loss} | test acc: {test_acc}")
```

I tried to help. i hope it works
great regards from
-shisirkhan
